distribution/pull_v2.go

```diff
@@ -75,7 +75,7 @@ type puller struct {
 
 func (p *puller) pull(ctx context.Context, ref reference.Named) (err error) {
 	// TODO(tiborvass): was ReceiveTimeout
-	p.repo, err = newRepository(ctx, p.repoInfo, p.endpoint, p.config.MetaHeaders, p.config.AuthConfig, "pull")
+	p.repo, err = NewRepository(ctx, p.repoInfo, p.endpoint, p.config.MetaHeaders, p.config.AuthConfig, "pull")
```

Comment:
I tried to avoid exposing too much, but I don't think we can unexpose more here. We can't even move it to internal, because we would need to reference private things from distribution, which we would need to make public anyway.

Reply:
Alternatively, I could make a function like `distribution.GetAllTags(registry.Service) ([]string, error)` so we'd avoid exposing `NewRepository` and `PullEndpoints`.

Reply:
Opened: https://github.com/moby/moby/pull/46618 as alternative